### PR TITLE
feat: add maintenance banner

### PR DESCRIPTION
**Pull request by brobro10000**
This pull request adds a system-wide maintenance warning banner to the application, which displays a configurable message to users when maintenance mode is enabled. The banner is conditionally shown based on environment variables and date checks, and leverages a new utility for determining alert visibility. The implementation also includes utility improvements for date handling.

Screenshot 2026-06-12 at 1 01 42 PM

**System maintenance alert feature:**
* Introduced a new `SystemWideWarningBanner` component that displays a warning banner with a customizable message and icon. (`src/components/SystemWideWarningBanner/SystemWideWarningBanner.tsx`)
* Added logic in `Layout.tsx` to display the warning banner when the maintenance alert is enabled and within the configured date range. (`src/components/app/Layout.tsx`)
* Added environment variables and configuration keys for maintenance alert enablement, message, and date range. (`src/index.tsx`)

**Utility and date handling improvements:**
* Added `isSystemMaintenanceAlertOpen` utility to encapsulate logic for determining if the maintenance alert should be shown, including checks for enablement, message, and date range. (`src/components/SystemWideWarningBanner/utils.ts`)
* Added `isTodayBetweenDates` utility and improved `dayjs` usage by centralizing plugin extensions in a new `dayjs.ts` file; updated imports to use this new module. (`src/utils/common.ts`, `src/utils/dayjs.ts`) [[1]](diffhunk://#diff-4ade25325a6cd7d1cca619364c84d19bfa47e6939a2c1b38d065767255b8f86eL4-R5) [[2]](diffhunk://#diff-4ade25325a6cd7d1cca619364c84d19bfa47e6939a2c1b38d065767255b8f86eR164-R170) [[3]](diffhunk://#diff-4ade25325a6cd7d1cca619364c84d19bfa47e6939a2c1b38d065767255b8f86eR182) [[4]](diffhunk://#diff-74b1779245c44578c598706ce7cd8ac2aa2300f739859889b164d5520fc83b28R1-R11)

### CODE REVIEW

File: src/components/app/Layout.tsx [hunk 1]
```diff
@@ -1,18 +1,29 @@
+import { getConfig } from '@edx/frontend-platform/config';
 import { Helmet } from 'react-helmet';
 import { Outlet, ScrollRestoration } from 'react-router-dom';
 
 import Header from '@/components/Header/Header';
+import SystemWideWarningBanner from '@/components/SystemWideWarningBanner/SystemWideWarningBanner';
+import { isSystemMaintenanceAlertOpen } from '@/components/SystemWideWarningBanner/utils';
```
[the remark this answers is not in the document]
Comment (pwnage101) [reply src/components/app/Layout.tsx]: yep, this seems legit, please do the module-boundary import refactor.

File: src/utils/common.ts [hunk 2]
```diff
@@ -160,6 +161,13 @@ const isFeatureEnabled = (enabled: boolean, featureKey?: string | null): boolean
   return isUnlockedBySiteKey || isUnlockedByFeatureKey;
 };
 
+function isTodayBetweenDates({ startDate, endDate }) {
+  const today = dayjs();
+  const formattedStartDate = dayjs(startDate);
+  const formattedEndDate = dayjs(endDate);
+  return today.isBetween(formattedStartDate, formattedEndDate);
```
Comment (pwnage101) [inline src/utils/common.ts]: There's a lot I think is misleading about this helper function:
* It is used to operate on datetimes, but its function name and argument names all suggest dates only.
* It sets variables called "formattedX" when the type is not a formatted string (they are Dayjs type).

The resulting re-write would just be one single line of code.  Please just delete this helper function and update callers to do this:

```typescript
dayjs().isBetween(dayjs(startTimestamp), dayjs(endTimestamp));
```

File: src/components/SystemWideWarningBanner/utils.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  const isEnabledWithMessage = (
+    config.IS_MAINTENANCE_ALERT_ENABLED && config.MAINTENANCE_ALERT_MESSAGE
+  );
+  if (!isEnabledWithMessage) {
+    return false;
+  }
+  const startTimestamp = config.MAINTENANCE_ALERT_START_TIMESTAMP;
+  const endTimestamp = config.MAINTENANCE_ALERT_END_TIMESTAMP;
+  if (startTimestamp && endTimestamp) {
+    return isTodayBetweenDates({ startDate: startTimestamp, endDate: endTimestamp });
```
Comment (pwnage101) [inline src/components/SystemWideWarningBanner/utils.ts]: Remove uselessly simple helper function:
```suggestion
    return dayjs().isBetween(dayjs(startTimestamp), dayjs(endTimestamp));
```

File: src/components/SystemWideWarningBanner/utils.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import dayjs from 'dayjs';
```
[the remark this answers is not in the document]
Comment (pwnage101) [reply src/components/SystemWideWarningBanner/utils.ts]: Do this so we can re-gain access to dayjs.isBetween().

Comment (pwnage101) [inline src/components/SystemWideWarningBanner/utils.ts]: I agree with copilot, do this to re-gain access to dayjs.isBetween():
```suggestion
import dayjs from '@/utils/dayjs';
```

File: src/components/SystemWideWarningBanner/utils.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ * @param {Object} config
+ * @returns {boolean}
+ */
+export function isSystemMaintenanceAlertOpen(config): boolean {
+  if (!config) {
+    return false;
+  }
+  const isEnabledWithMessage = (
+    config.IS_MAINTENANCE_ALERT_ENABLED && config.MAINTENANCE_ALERT_MESSAGE
+  );
```
[the remark this answers is not in the document]
Comment (pwnage101) [reply src/components/SystemWideWarningBanner/utils.ts:14]: This is interesting.  But this bug shouldn't manifest in stage/prod because we pass yaml files directly which are ingested by mergeConfig() as boolean type.  Maybe boolean settings don't work when loaded from environment files (`.env.*`)?  Since this doesn't impact stage/prod, and fixing it should just fix ALL boolean settings, lets just ignore this for now to keep the scope narrow.

**Review by pwnage101 — CHANGES_REQUESTED** [review]

File: src/utils/common.ts (hunk 2 above)
Comment (brobro10000) [reply src/utils/common.ts]: `dayjs` is able to parse the format of the date and normalize it into a dayjs object regardless whether its a date, or timestamp.

Perhaps this could be returned as `normalizedStartDate` and `normalizeEndDate`.

This function still has value since it no longer requires downstream consumers to use this method and pass in 2 further normalized dayjs objects.

File: src/components/SystemWideWarningBanner/utils.ts (hunk 3 above)
Comment (brobro10000) [reply src/components/SystemWideWarningBanner/utils.ts]: I would argue to function since it simplifies the overall approach to a single function, encapsulating date normalization within the function itself, versus having the consumer having to do it for each value. 

The encapsulation is not useless and serves a purpose, similar to this 1 liner in python.
https://github.com/edx/enterprise-access/blob/970dd43d526123d9aa579999c525cac1e9b8af95/enterprise_access/apps/customer_billing/utils.py

Comment (pwnage101) [reply src/components/SystemWideWarningBanner/utils.ts]: The purpose of a function is to abstract business-specific logic.  checking if an input date is between two other dates is completely accomplished with isBetween and we are not calling it in a weird way or deviating from the dayjs docs in any way.

By your logic, `dayjs().isAfter(dayjs(startTimestamp));` should also become a one-line function, as well as `dayjs().isBefore(dayjs(endTimestamp));`. 

Furthermore, I'd argue the linked python function also has little utility.

Comment (brobro10000) [reply src/components/SystemWideWarningBanner/utils.ts]: Agree to disagree. refactored.

File: src/components/SystemWideWarningBanner/utils.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  const endTimestamp = config.MAINTENANCE_ALERT_END_TIMESTAMP;
+  if (startTimestamp && endTimestamp) {
+    return dayjs().isBetween(dayjs(startTimestamp), dayjs(endTimestamp));
+  }
+  if (startTimestamp) {
+    return dayjs().isAfter(dayjs(startTimestamp));
+  }
+  if (endTimestamp) {
+    return dayjs().isBefore(dayjs(endTimestamp));
+  }
```
[the remark this answers is not in the document]
Comment (pwnage101) [reply src/components/SystemWideWarningBanner/utils.ts:28]: again, we're not worried about inclusivity of the boundaries which would have no perceivable difference in practice.

File: src/components/SystemWideWarningBanner/utils.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * Check if system maintenance alert is open, based on configuration.
+ * @param {Object} config
+ * @returns {boolean}
+ */
+export function isSystemMaintenanceAlertOpen(config): boolean {
+  if (!config) {
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (pwnage101) [reply src/components/SystemWideWarningBanner/utils.ts:11]: Not conventional, would be nice but this is not the solution (there could be other keys).

File: src/components/SystemWideWarningBanner/tests/utils.test.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    {
+      scenario: 'now is outside start/end window',
+      config: {
+        IS_MAINTENANCE_ALERT_ENABLED: true,
+        MAINTENANCE_ALERT_MESSAGE: 'Message',
+        MAINTENANCE_ALERT_START_TIMESTAMP: dayjs(fixedNow).add(1, 'hour').toISOString(),
+        MAINTENANCE_ALERT_END_TIMESTAMP: dayjs(fixedNow).add(2, 'hour').toISOString(),
+      },
+      expected: false,
+    },
```
[the remark this answers is not in the document]
Comment (pwnage101) [reply src/components/SystemWideWarningBanner/tests/utils.test.ts:66]: For the first time ever, I don't care about boundary conditions.  They can be undefined behavior, and that's okay.

File: src/components/app/tests/Layout.test.tsx [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+      </MemoryRouter>
+    </IntlProvider>,
+  );
+};
+
+describe('Layout', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+    jest.spyOn(console, 'log').mockImplementation(() => {});
+  });
```
[the remark this answers is not in the document]
Comment (pwnage101) [reply src/components/app/tests/Layout.test.tsx:56]: Seems legit, why mock something that doesn't need to be mocked.

[the remark this answers is not in the document]
Comment (pwnage101) [reply src/components/app/tests/Layout.test.tsx:56]: duplicate

[the remark this answers is not in the document]
Comment (pwnage101) [reply src/components/app/tests/Layout.test.tsx:56]: duplicate

File: src/components/SystemWideWarningBanner/utils.ts (hunk 7 above)
Comment (pwnage101) [inline src/components/SystemWideWarningBanner/utils.ts:11]: Nit: why not just call getConfig() from this function instead of passing it via argument?  getConfig isn't a hook, so it should be safe to be call from other non-hook utility functions.  It's like in django we always just import settings and use `settings.FOOBAR_ENALBED` wherever we need it, but we never actually pass the `settings` object around.

**Review by pwnage101 — APPROVED** [review]
